src/vos/vos_aggregate.c

```diff
@@ -589,6 +672,150 @@ merge_window_size(struct agg_merge_window *mw)
 	return evt_extent_width(&mw->mw_ext) * mw->mw_rsize;
 }
 
+/* Widen biov entry for read extents to range required to verify checksums. */
+static unsigned int
+csum_widen_biov(struct bio_iov *biov, struct agg_phy_ent *phy_ent,
+		    struct evt_extent *ext, uint32_t rsize, daos_off_t phy_lo,
+		    uint32_t *wider)
+{
+	struct evt_entry	ent;
+	struct evt_extent	aligned_extent = { 0 };
+	unsigned int		added_segs = 0;
+
+	ent.en_ext = phy_ent->pe_rect.rc_ex;
+	if (phy_lo)
+		ent.en_ext.ex_lo = phy_lo;
+	ent.en_sel_ext = *ext;
+	ent.en_csum = phy_ent->pe_csum_info;
+	aligned_extent = evt_entry_align_to_csum_chunk(&ent, rsize);
+	bio_iov_set_extra(biov,
+			  (ent.en_sel_ext.ex_lo - aligned_extent.ex_lo) *
+			  rsize,
+			  (aligned_extent.ex_hi - ent.en_sel_ext.ex_hi) *
+			  rsize);
+	/*Amount to add to IO buffer. */
+	*wider = biov->bi_prefix_len + biov->bi_suffix_len;
+	added_segs += biov->bi_prefix_len != 0;
+	added_segs += biov->bi_suffix_len != 0;
+	/* Number of additional read segments for this component extent. */
+	return added_segs;
+}
+
+
+/* Extends bio_sglist to include extension to csum boumdaries (added to the end
+ * of the list) with added_segs additional entries. These entries will hold the
+ * extended prefix and suffix data ranges needed to widen the aggregatable range
+ * to prior checksum boundaries.
+ */
```

Comment:
I'm bit lost here, could you explain how these appended biovs are used?

Reply:
These are the added prefix/suffix ranges that are needed to calculate the checksum for the component (input) extents.  They are added at the end to allow a contiguous range of the write-able ranges (without the added prefix/suffix ranges) in the initial range of the buffer.  This initial range is then passed in the bio_write call.

Reply:
I got it. Thanks.